### DISCUSSION TITLE
Performance difference between SymbolTypes

### DISCUSSION THREAD

**Discussion by james-s-may** [Mapsui/Mapsui#2432] [Q&A]
Hi there,

I have a project where I have many features, each with multiple stacked styles. Each has an image/bitmap symbol style, plus a rectangle symbol style (an icon plus a background rectangle).

I have created a separate project to compare the performance of the image styles vs rectangle styles. It seems that thousands of rectangle symbols runs significantly slower than thousands of image symbols.

Why is this, and is there any tips on how to increase the performance of SymbolStyles, or another method? Any advice is appreciated!

I am using Avalonia, with latest release version of both Avalonia and Mapsui.

Steps to reproduce my test project:

1. Create a new Avalonia project with Mapsui, following:  https://mapsui.com/documentation/getting-started-avalonia.html
2. Paste the below code into MainView.axaml.cs
3. Replace the project name in the namespace on line 15 and the path on line 81 with your project name
4. Change the _useImage field to compare. Rebuild and run in browser for most dramatic performance change. I get ~70fps with images, and ~12fps with rectangles

```
using System;
using System.Collections.Generic;
using Avalonia.Controls;
using Avalonia.Platform;
using Mapsui;
using Mapsui.Extensions;
using Mapsui.Layers;
using Mapsui.Projections;
using Mapsui.Rendering.Skia.SkiaWidgets;
using Mapsui.Styles;
using Mapsui.UI.Avalonia;
using Mapsui.Widgets;
using Mapsui.Widgets.PerformanceWidget;

namespace AvaloniaApplication4.Views;

public partial class MainView : UserControl
{
    
    private bool _useImage = false;
    private int _symbolCount = 1000;
    
    
    public MainView()
    {
        InitializeComponent();
        
        var mapControl = new Mapsui.UI.Avalonia.MapControl();
        mapControl.Map?.Layers.Add(Mapsui.Tiling.OpenStreetMap.CreateTileLayer());
        Content = mapControl;
        
        mapControl.Map?.Layers.Add(CreateSymbolsLayer());

        CreatePerformanceWidget(mapControl);
    }


    private void CreatePerformanceWidget(MapControl map)
    {
        map.Performance ??= new Mapsui.Utilities.Performance();

        void WidgetTouched(object? sender, WidgetTouchedEventArgs args)
        {
            map.Performance.Clear();
            map.RefreshGraphics();

            args.Handled = true;
        };

        var performanceWidget = new PerformanceWidget(map.Performance);
        performanceWidget.WidgetTouched += WidgetTouched;
        map.Map.Widgets.Add(performanceWidget);
        map.Renderer.WidgetRenders[typeof(PerformanceWidget)] = new PerformanceWidgetRenderer(
            10, 10, 12, SkiaSharp.SKColors.Black, SkiaSharp.SKColors.White);
    }
    
    private ILayer CreateSymbolsLayer()
    {
        return new MemoryLayer
        {
            Name = "Symbols Layer",
            Features = CreateFeatures(),
            Style = null,
            IsMapInfoLayer = true
        };
    }
    
    private IEnumerable<IFeature> CreateFeatures()
    {
        var features = new List<IFeature>();
        
        double latBase = 51.5;
        double latRange = 0.2;
        double lonBase = -2.5;
        double lonRange = 0.4;
        Random rand = new Random();
        
        SymbolStyle style;
        if (_useImage)
        {
            var stream = AssetLoader.Open(new Uri("avares://AvaloniaApplication4/Assets/avalonia-logo.ico"));
            int bitmapId = BitmapRegistry.Instance.Register(stream);
            style = new SymbolStyle { SymbolType = SymbolType.Image, BitmapId = bitmapId, SymbolScale = 0.1 };
        }
        else
        {
            style = new SymbolStyle { SymbolType = SymbolType.Rectangle };
        }
        
        for (int i = 0; i < _symbolCount; i++)
        {
            double latOffset = (rand.NextDouble() * (latRange * 2)) - latRange;
            double lonOffset = (rand.NextDouble() * (lonRange * 2)) - lonRange;
            double lat = latBase + latOffset;
            double lon = lonBase + lonOffset;
            MPoint mapPos = SphericalMercator.FromLonLat(lon, lat).ToMPoint();

            var feature = new PointFeature(mapPos);
            feature.Styles.Add(style);
            features.Add(feature);

        }
        return features;
    }
    
}
```

**Comment by pauldendulk** [comment]
Interesting, drawing a rectangle seems super simple. I could imagine that drawing images uses soms hardware acceleration, but 12fps for 1000 simple rectangles is really low. I would think something weird is going on. Glancing at the code I just see a simple SkiaSharp draw method. I don't expect it to be this slow (but perhaps it is, and we should cache it into an SKBitmap).

With my current priorities, this is not something I am going to look into soon. Perhaps someone else could investigate this. You do not need to much specific knowledge. SkiaSharp knowledge would help. You need to clone the Mapsui project, add the sample as described above and search for reference of SymbolType.Rectangle to find the drawing code.

**Comment by charlenni** [comment]
I looked into this and get on the desktop the expected behavior: the rectangles are a little faster than the images. Around 40 fps against 35 fps. I looked also in the code and could find any unusual.
